File: cpp_version/Swiat.cpp
#include "Swiat.h"
#include "Organizm.h"
#include <conio.h>
#include <fstream>
#include <sstream>
#include "Lis.h"
#include "Antylopa.h"
#include "BarszczSosnowskiego.h"
#include "Cyberowca.h"
#include "Czlowiek.h"
#include "Guarana.h"
#include "Mlecz.h"
#include "Owca.h"
#include "Trawa.h"
#include "WilczeJagody.h"
#include "Wilk.h"
#include "Zolw.h"

Swiat::Swiat(int szerokosc, int wysokosc) : szerokosc(szerokosc), wysokosc(wysokosc) {
    plansza.resize(wysokosc, std::vector<char>(szerokosc, '.'));
}

void Swiat::dodajOrganizm(std::shared_ptr<Organizm> organizm) {
	if (organizm == nullptr) {
		std::cerr << "Nie mozna dodac pustego organizmu!" << std::endl;
		return;
	}
    organizmy.push_back(organizm);
    plansza[organizm->getY()][organizm->getX()] = organizm->rysowanie();
    wiekOrganizmow[organizm.get()] = 0;
    dodajLog("Dodano " + organizm->getNazwa() + " na pole (" + std::to_string(organizm->getX()) + ", " + std::to_string(organizm->getY()) + ")");
    organizm->nowoNarodzony = true;
}

void Swiat::wykonajTure() {
    for (auto& org : organizmy) {
        wiekOrganizmow[org.get()]++;
    }

    std::sort(organizmy.begin(), organizmy.end(), [this](const std::shared_ptr<Organizm>& a, const std::shared_ptr<Organizm>& b) {
        if (a->getInicjatywa() != b->getInicjatywa())
            return a->getInicjatywa() > b->getInicjatywa();
        return wiekOrganizmow[a.get()] > wiekOrganizmow[b.get()];
        });

    std::vector<std::shared_ptr<Organizm>> kopiaOrganizmow = organizmy;

    for (auto& organizm : kopiaOrganizmow) {
        if (organizm->czyZyje() && !organizm->nowoNarodzony) {
            organizm->akcja();
        }
    }

    if (organizmy.size() > (getSzerokosc()*getWysokosc()*2)) {
        std::cerr << "ZA DUZO ORGANIZMOW!\n";
        exit(1);
    }

    usunMartweOrganizmy();
    odswiezPlansze();
    rysujSwiat();
    wyswietlLogi();
    logi.clear();
    for (auto& org : organizmy) {
        org->nowoNarodzony = false;
    }
    return;
}


void Swiat::rysujSwiat() {
    std::cout << "\n=== PLANSZA ===\n";
    for (const auto& wiersz : plansza) {
        for (const auto& pole : wiersz) {
            std::cout << pole << " ";
        }
        std::cout << "\n";
    }
}

void Swiat::inicjalizujPlansze() {
    plansza.clear(); // wyczysc plansze
    plansza.resize(wysokosc, std::vector<char>(szerokosc, '.'));
}


Organizm* Swiat::znajdzNajblizszyBarszcz(int x, int y) const {
    Organizm* najblizszy = nullptr;
    int minDystans = INT_MAX;
    for (const auto& org : organizmy) {
        if (org->czyZyje() && org->getNazwa() == "BarszczSosnowskiego") {
            int dx = x - org->getX();
            int dy = y - org->getY();
            int dystans = abs(dx) + abs(dy);

            if (dystans < minDystans) {
                minDystans = dystans;
                najblizszy = org.get();
            }
        }
    }
    return najblizszy;
}


void Swiat::dodajLog(const std::string& log) {
    logi.push_back(log);
}

void Swiat::wyswietlLogi() const {
    std::cout << "\n=== LOGI ===\n";
    for (const auto& log : logi) {
        std::cout << log << "\n";
    }
}

int Swiat::getSzerokosc() const {
    return szerokosc;
}

int Swiat::getWysokosc() const {
    return wysokosc;
}

void Swiat::aktualizujPole(int x, int y, char symbol) {
    plansza[y][x] = symbol;
}

bool Swiat::czyPolePuste(int x, int y) const {
    return plansza[y][x] == '.';
}

Organizm* Swiat::getOrganizm(int x, int y) const {
    for (const auto& org : organizmy) {
        if (org->getX() == x && org->getY() == y && org->czyZyje()) {
            return org.get();
        }
    }
    return nullptr;
}


void Swiat::usunMartweOrganizmy() {
    size_t przed = organizmy.size();
    organizmy.erase(
        std::remove_if(organizmy.begin(), organizmy.end(),
            [](const std::shared_ptr<Organizm>& org) {
                return !org->czyZyje();
            }),
        organizmy.end());
    size_t po = organizmy.size();
    if (przed != po) {
        dodajLog("Usunietych organizmow: " + std::to_string(przed - po));
    }
}

void Swiat::odswiezPlansze() {
    for (auto& wiersz : plansza)
        std::fill(wiersz.begin(), wiersz.end(), '.');

    for (const auto& org : organizmy) {
        plansza[org->getY()][org->getX()] = org->rysowanie();
    }
}

void Swiat::zapiszDoPliku(const std::string& nazwaPliku) {

    std::ofstream plik(nazwaPliku);

    if (!plik.is_open()) {
        std::cerr << "Nie mozna otworzyc pliku do zapisu.\n";
        return;
    }

    plik << getSzerokosc() << " " << getWysokosc() << "\n";

    for (auto& organizm : organizmy) {
        if (organizm->czyZyje()) {
            plik << organizm->getNazwa() << " "
                << organizm->getSila() << " "
                << organizm->getX() << " "
                << organizm->getY();

            if (Czlowiek* czlowiek = dynamic_cast<Czlowiek*>(organizm.get())) {
                plik << " cooldown:" << czlowiek->getCooldown()
                    << " aktywna:" << czlowiek->isAktywnaUmiejetosc()
                    << " czasTrwania:" << czlowiek->getCzasTrwania();
            }

            plik << "\n";
        }
    }

    plik.close();
    std::cout << "[Swiat] Stan gry zapisany do pliku: " << nazwaPliku << "\n";
}

void Swiat::wczytajZPliku(const std::string& nazwaPliku) {
    std::ifstream plik(nazwaPliku);

    if (!plik.is_open()) {
        std::cerr << "Nie mozna otworzyc pliku do odczytu.\n";
        return;
    }

    organizmy.clear();
    plansza.clear();

    plik >> szerokosc >> wysokosc;
    inicjalizujPlansze();

    std::string linia;
    std::getline(plik, linia);

    while (std::getline(plik, linia)) {
        if (linia.empty()) continue;

        std::istringstream iss(linia);
        std::string nazwa;
        int sila, x, y;

        if (!(iss >> nazwa >> sila >> x >> y)) {
            std::cerr << "Blad formatu pliku: " << linia << std::endl;
            continue;
        }

        std::shared_ptr<Organizm> organizm = nullptr;

        if (nazwa == "Lis") organizm = std::make_shared<Lis>(x, y, *this);
        else if (nazwa == "Owca") organizm = std::make_shared<Owca>(x, y, *this);
        else if (nazwa == "Trawa") organizm = std::make_shared<Trawa>(x, y, *this);
        else if (nazwa == "Guarana") organizm = std::make_shared<Guarana>(x, y, *this);
        else if (nazwa == "Wilk") organizm = std::make_shared<Wilk>(x, y, *this);
        else if (nazwa == "Zolw") organizm = std::make_shared<Zolw>(x, y, *this);
        else if (nazwa == "Antylopa") organizm = std::make_shared<Antylopa>(x, y, *this);
        else if (nazwa == "CyberOwca") organizm = std::make_shared<CyberOwca>(x, y, *this);
        else if (nazwa == "Mlecz") organizm = std::make_shared<Mlecz>(x, y, *this);
        else if (nazwa == "WilczeJagody") organizm = std::make_shared<WilczeJagody>(x, y, *this);
        else if (nazwa == "BarszczSosnowskiego") organizm = std::make_shared<BarszczSosnowskiego>(x, y, *this);
        else if (nazwa == "Czlowiek") {
            auto czlowiek = std::make_shared<Czlowiek>(x, y, *this);

            std::string klucz;
            while (iss >> klucz) {
                if (klucz.find("cooldown:") == 0)
                    czlowiek->setCooldown(std::stoi(klucz.substr(9)));
                else if (klucz.find("aktywna:") == 0)
                    czlowiek->setAktywnaUmiejetnosc(std::stoi(klucz.substr(8)));
                else if (klucz.find("czasTrwania:") == 0)
                    czlowiek->setCzasTrwania(std::stoi(klucz.substr(12)));
            }

            organizm = czlowiek;
        }

        if (organizm != nullptr) {
            organizm->zwiekszSile(sila - organizm->getSila());
            dodajOrganizm(organizm);
        }
    }

    plik.close();
    std::cout << "[Swiat] Stan gry wczytany z pliku: " << nazwaPliku << "\n";
}

File: python_version/Swiat.py
from Czlowiek import Czlowiek
from Wilk import Wilk
from Trawa import Trawa
# + inne do dodania później
import random

class Swiat:
    def __init__(self, szerokosc, wysokosc):
        self.szerokosc = szerokosc
        self.wysokosc = wysokosc
        self.plansza = [['.' for _ in range(szerokosc)] for _ in range(wysokosc)]
        self.organizmy = []
        self.logi = []
        self.wiek = {}

    def dodaj_organizm(self, organizm):
        if organizm is None: return
        self.organizmy.append(organizm)
        self.plansza[organizm.get_y()][organizm.get_x()] = organizm.rysowanie()
        self.wiek[organizm] = 0
        organizm.nowo_narodzony = True
        self.dodaj_log(f"Dodano {organizm.get_nazwa()} na ({organizm.get_x()}, {organizm.get_y()})")

    def wykonaj_ture(self):
        for org in self.organizmy:
            self.wiek[org] += 1

        self.organizmy.sort(key=lambda o: (-o.get_inicjatywa(), -self.wiek[o]))

        for org in list(self.organizmy):
            if org.czy_zyje() and not org.nowo_narodzony:
                org.akcja()

        self.usun_martwe()
        self.odswiez_plansze()
        #self.rysuj()
        #self.wyswietl_logi()
        #self.logi.clear()
        for org in self.organizmy:
            org.nowo_narodzony = False

    def odswiez_plansze(self):
        self.plansza = [['.' for _ in range(self.szerokosc)] for _ in range(self.wysokosc)]
        for org in self.organizmy:
            if org.czy_zyje():
                self.plansza[org.get_y()][org.get_x()] = org.rysowanie()

    def rysuj(self):
        print("=== PLANSZA ===")
        #for w in self.plansza:
        #    print(' '.join(w))

    def dodaj_log(self, txt):
        self.logi.append(txt)

    def wyswietl_logi(self):
        print("=== LOGI ===")
        #for log in self.logi:
        #    print(log)

    def get_szerokosc(self):
        return self.szerokosc

    def get_wysokosc(self):
        return self.wysokosc

    def aktualizuj_pole(self, x, y, symbol):
        self.plansza[y][x] = symbol

    def czy_pole_puste(self, x, y):
        if 0 <= x < self.szerokosc and 0 <= y < self.wysokosc:
            return self.plansza[y][x] == '.'
        return False

    def get_organizm(self, x, y):
        for org in self.organizmy:
            if org.get_x() == x and org.get_y() == y and org.czy_zyje():
                return org
        return None

    def get_organizmy(self):
        return self.organizmy

    def usun_martwe(self):
        przed = len(self.organizmy)
        self.organizmy = [org for org in self.organizmy if org.czy_zyje()]
        po = len(self.organizmy)
        if przed != po:
            self.dodaj_log(f"Usunięto {przed - po} martwych organizmów")

    def znajdz_najblizszy_barszcz(self, x, y):
        najblizszy = None
        min_dystans = float('inf')
        for org in self.organizmy:
            if org.czy_zyje() and org.get_nazwa() == "BarszczSosnowskiego":
                dystans = abs(x - org.get_x()) + abs(y - org.get_y())
                if dystans < min_dystans:
                    min_dystans = dystans
                    najblizszy = org
        return najblizszy

